File: mobile/Chela/OutboxReconcile.swift
import Foundation

/// The outbox reconcile, read from `core/spec/outbox-reconcile.json`.
///
/// The Control UI rewrites every in-flight row of its outbox as if the page had
/// reloaded, on every pass through its codec. A message the gateway already
/// accepted is then shown as "Reconnected before delivery was confirmed" on a
/// connection that never dropped, and the messages behind it can wait on a
/// reconnect that never comes. The script checks such a row against the gateway's
/// own record over the page's own socket: a row the gateway holds is removed as
/// delivered, and a row it does not hold is sent once, under its own send id, so
/// the gateway's dedupe keeps it to one copy.
///
/// The script is not ported: it is read from the spec the app bundles, exactly as
/// `ReconnectResumeShim` reads its hook, so the bytes this client installs are the
/// bytes `core/outbox-reconcile.js` hands the desktop. It is installed at document
/// START because it wraps the page's `WebSocket` constructor, which has to happen
/// before the page opens its socket. The spec's own `why` list names the upstream
/// fixes that retire it.
enum OutboxReconcile {
    private struct Spec: Decodable {
        let global: String
        let hook: [String]
    }

    private static let spec: Spec = loadSpec()

    private static func loadSpec() -> Spec {
        let empty = Spec(global: "", hook: [])
        guard let url = Bundle.main.url(forResource: "outbox-reconcile", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let spec = try? JSONDecoder().decode(Spec.self, from: data),
              !spec.global.isEmpty,
              !spec.hook.isEmpty
        else {
            // A build that did not bundle the spec installs nothing rather than
            // something invented, and `OutboxReconcileParityTests` is what turns
            // that into a failing build rather than a quiet absence in the field.
            return empty
        }
        return spec
    }

    /// The install-guard global the script sets on the page, which also carries its log.
    static var global: String { spec.global }

    /// The script, exactly as `core/spec/outbox-reconcile.json` holds it.
    static var script: String { spec.hook.joined(separator: "\n") }
}

File: core/outbox-reconcile.js
// The outbox reconcile: the document-start script both clients install so a
// message the Control UI queued is never left stuck on a live connection.
//
// The Control UI keeps its outbox in sessionStorage and, on every pass through
// its codec, rewrites an in-flight row as if the page had reloaded. A send the
// gateway already accepted then reads "Reconnected before delivery was
// confirmed" on a socket that never dropped, and the rows behind it wait on a
// reconnect that never comes. The script checks such a row against the gateway's
// own record over the page's own socket, removes it when the gateway holds it,
// and sends it once, under its own send id, when the gateway does not. The why,
// the thresholds and the removal trigger are in spec/outbox-reconcile.json.
//
// Platform-free and data-driven, like the reconnect-resume shim: the script
// lives in the spec, the desktop evaluates it from its preload and the phone
// installs it through a WKUserScript, so there is one copy of it in the tree.

import spec from './spec/outbox-reconcile.json' with { type: 'json' };

/** The install-guard global, which also carries the script's state and log for a reader. */
export const RECONCILE_GLOBAL = spec.global;

/**
 * The injected script, exactly as the spec holds it.
 *
 * @returns {string} the script, one line per spec line
 */
export function reconcileScript() {
  return spec.hook.join('\n');
}
